lib/atat-web-api-stack.ts

```diff
@@ -1,11 +1,85 @@
+import * as lambdaNodeJs from "aws-cdk-lib/aws-lambda-nodejs";
+import {
+  StateMachine,
+  maxRetries,
+  successResponse,
+  clientErrorResponse,
+  internalErrorResponse,
+} from "./constructs/stateMachine";
+import { mapTasks } from "./constructs/sfnLambdaInvokeTask";
 import { AtatRestApi } from "./constructs/apigateway";
+import { ApiSfnFunction } from "./constructs/ApiSfnFunction";
+import { HttpMethod } from "../utils/http";
 
 export class AtatWebApiStack extends Stack {
+  public readonly provisioningStateMachine: sfn.IStateMachine;
+
   constructor(scope: Construct, id: string, props?: StackProps) {
     super(scope, id, props);
     const apigw = new AtatRestApi(this, "SampleApi").restApi;
     apigw.root.addMethod("ANY");
+    // TODO: add startProvisionJob lambda to API GW
+
+    // Provisioning State machine functions
+    const sampleFn = new lambdaNodeJs.NodejsFunction(this, "SampleFunction", {
+      entry: "api/provision/sampleFn.ts",
+      // vpc: props.vpc,
+    });
+
+    const tasks = [
+      {
+        id: "InvokeCspApi",
+        props: {
+          lambdaFunction: sampleFn,
+          inputPath: "$",
+          resultSelector: {
+            statusCode: sfn.JsonPath.stringAt("$.Payload.passRequest.response.statusCode"),
+            body: sfn.JsonPath.stringAt("$.Payload.body"),
+            retryCount: sfn.JsonPath.stringAt("$$.State.RetryCount"),
+          },
+          resultPath: "$.cspResponse",
+          outputPath: "$",
+        },
+      },
+      {
+        id: "RejectPortfolioTask",
+        props: {
+          lambdaFunction: sampleFn,
+          inputPath: "$.cspResponse",
+        },
+      },
+    ];
+    const mappedTasks = mapTasks(this, tasks);
+
+    // State machine choices
+    const httpResponseChoices = new sfn.Choice(this, "HttpResponse")
+      .when(successResponse, mappedTasks.InvokeCspApi.sfnTask)
+      .when(clientErrorResponse, mappedTasks.RejectPortfolioTask.sfnTask)
+      .when(maxRetries, mappedTasks.RejectPortfolioTask.sfnTask)
+      .when(internalErrorResponse, mappedTasks.RejectPortfolioTask.sfnTask);
+
+    // Composing state machine
+    const stateMachineWorkflow = mappedTasks.InvokeCspApi.sfnTask.next(httpResponseChoices);
+    const stateMachineLogGroup = new logs.LogGroup(this, "StepFunctionsLogs", {
+      retention: logs.RetentionDays.ONE_WEEK,
+      logGroupName: `/aws/vendedlogs/states/StepFunctionsLogs${"AT-7050"}`,
+    });
+    this.provisioningStateMachine = new StateMachine(this, "ProvisioningStateMachine", {
+      stateMachineProps: {
+        definition: stateMachineWorkflow,
+      },
+      logGroup: stateMachineLogGroup,
+    }).stateMachine;
+
+    // the provisioning lambda that translate and invokes the state machine
+    const provisioningJobFn = new ApiSfnFunction(this, "ProvisioningJobRequest", {
+      method: HttpMethod.POST,
+      handlerPath: "api/provision/startProvisionJob.ts",
+      stateMachine: this.provisioningStateMachine,
+    }).fn;
```

Comment:
This all feels fairly complicated to be at the top-level of the stack -- can we abstract it into its own construct in a new file?

Reply:
I agree and tried to trim it down, but I'll try to abstract it await into its own construct.